### PR TITLE
[FLINK-16513][checkpointing] Unaligned checkpoints: checkpoint metadata

### PR DESCRIPTION
**Pull request by rkhachatryan**
## What is the purpose of the change

Currently, Flink only snapshots and stores operator state. For the Unaligned Checkpoints, channel state needs to be stored as well. This PR adds channel state handles to checkpoint metadata and updates state assignment on recovery.

The metadata is added on the `OperatorSubtaskState` level: each operator has in and out channel states. Currently, only head and tail operators in the chain will have `in` and `out` states respectively (topology change is not supported).
An alternative approach (implemented [here](https://github.com/rkhachatryan/flink/tree/uc-state-metadata)) would be to add channel state on the `TaskStateSnapshot` level. While it looks simpler from the Task perspective; it requires some artificial binding in `CompletedCheckpoint` to allow assignment on recovery.

## Brief change log

 - some pre-requisite refactorings
 - add channel state metadata to OperatorSubtaskState
 - adjust StateAssignmentOperation
 - adjust metadata serialization
 - adjust related classes

## Verifying this change

This change is already covered by existing (modified) tests, such as:
 - CheckpointCoordinatorFailureTest
 - LocalStateForwardingTest
 - OperatorSnapshotFinalizerTest
 - OperatorSnapshotFuturesTest
 - RestoreStreamTaskTest
 - StreamTaskTest

And new tests:
- `ChannelStateNoRescalingPartitionerTest`
- `StateAssignmentOperationTest#testChannelStateAssignmentStability`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no (not user-facing)
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV3Serializer.java [hunk 1]
```diff
@@ -115,4 +120,31 @@ protected OperatorState deserializeOperatorState(DataInputStream dis) throws IOE
...
+	@VisibleForTesting
+	@Override
+	public void serializeResultSubpartitionStateHandle(ResultSubpartitionStateHandle handle, DataOutputStream dos) throws IOException {
+		channelStateHandleSerializer.serialize(handle, dos);
+	}
+
+	@VisibleForTesting
+	@Override
+	public ResultSubpartitionStateHandle deserializeResultSubpartitionStateHandle(DataInputStream dis) throws IOException {
+		final boolean hasResultSubpartitionStateHandle = dis.readInt() != 0;
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV3Serializer.java]: are you sure that `dis.readInt() != 0` is good enough check to maintain backward compatibility?

Do we have test coverage for that?

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorFailureTest.java [hunk 2]
```diff
@@ -116,6 +122,8 @@ public void testFailingCompletedCheckpointStoreAdd() throws Exception {
 		verify(operatorSubtaskState.getRawOperatorState().iterator().next()).discardState();
 		verify(operatorSubtaskState.getManagedKeyedState().iterator().next()).discardState();
 		verify(operatorSubtaskState.getRawKeyedState().iterator().next()).discardState();
+		verify(operatorSubtaskState.getInputChannelState().iterator().next()).discardState();
+		verify(operatorSubtaskState.getResultSubpartitionState().iterator().next()).discardState();
```
[the remark this answers is not in the document]
Comment (rkhachatryan) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorFailureTest.java]: Agree, but this is an existing test which already uses mockito.
Having some fields mocked and some not would be confusing and changing the whole test is out of scope of this PR.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV3Serializer.java (hunk 1 above)
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV3Serializer.java]: Backward compatibility is maintained on a snapshot level (by having version number at the beginning of it). This check is for cases when there is no state (same approach as for other parts of the state).

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/OperatorSnapshotFutures.java [hunk 3]
```diff
@@ -105,41 +126,51 @@ public void setOperatorStateRawFuture(
...
+					}
+				});
+			}
 		}
+	}
 
-		if (exception != null) {
-			throw exception;
-		}
+	public Future<?>[] getAllFutures() {
```
Comment (Myasuka) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/OperatorSnapshotFutures.java]: Why not return `inputChannelStateFuture` and `resultSubpartitionStateFuture`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StateAssignmentOperation.java [hunk 4]
```diff
@@ -640,4 +639,17 @@ private static void checkStateMappingCompleteness(
 
 		return subtaskKeyedStateHandles;
 	}
+
+	private static <T extends AbstractChannelStateHandle<?>> OperatorStateRepartitioner<T> nonRescalingPartitioner(String logStateName) {
```
Comment (Myasuka) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StateAssignmentOperation.java]: I think there is no UT or IT to cover this method currently in this PR (e.g. throw expected exception).

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorFailureTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorFailureTest.java]: I'm not asking to rewrite the whole test (although that would be nice ;) ). Having said that, I would prefer to have mixed test instead of increasing our dependency on the mockito, as this is adding more and more technological debt. 

Also we have lot's of existing tests that are mixing mockito mocked classes with proper classes, so I don't think this is an issue.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV3Serializer.java (hunk 1 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV3Serializer.java]: Ok. 

Do we have proper test coverage for backward compatibility?

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorFailureTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rkhachatryan) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorFailureTest.java]: Yes, I do have strong opposite feeling :) 
(I see only 2 new lines; and disadvantages of mixed tests).

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StateAssignmentOperation.java (hunk 4 above)
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StateAssignmentOperation.java]: That's right. I'm adding some new tests including `ChannelStateNoRescalingPartitionerTest`.'

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV3Serializer.java (hunk 1 above)
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV3Serializer.java]: Just to clarify:
- V3 serializer was added in this release (1.11); initially, it was added for FLIP-27 operator coordinators state
- the versioning mechanism was added several years ago

So any test that reads a snapshot from file covers this because these files were written with serializer V2 or older. There are a lot of such tests.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/ChannelStateHandleSerializer.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+class ChannelStateHandleSerializer {
+
+	public void serialize(ResultSubpartitionStateHandle handle, DataOutputStream dos) throws IOException {
+		serializeChannelStateHandle(handle, dos, (info, dataOutputStream) -> {
+			dos.writeInt(info.getPartitionIdx());
+			dos.writeInt(info.getSubPartitionIdx());
+		});
+	}
+
+	ResultSubpartitionStateHandle deserializeResultSubpartitionStateHandle(DataInputStream dis) throws IOException {
```
[the remark this answers is not in the document]
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/ChannelStateHandleSerializer.java]: I'd rather choose stick to convention.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/ChannelStateHandleSerializer.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.flink.runtime.checkpoint.metadata.MetadataV2V3SerializerBase.deserializeStreamStateHandle;
+import static org.apache.flink.runtime.checkpoint.metadata.MetadataV2V3SerializerBase.serializeStreamStateHandle;
+
+class ChannelStateHandleSerializer {
+
+	public void serialize(ResultSubpartitionStateHandle handle, DataOutputStream dos) throws IOException {
```
[the remark this answers is not in the document]
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/ChannelStateHandleSerializer.java]: I'd rather choose stick to convention.

**Comment by rkhachatryan** [comment]
Thanks for your feedback @pnowojski and @Myasuka !
I've updated the PR, please take a look.

**Comment by rkhachatryan** [comment]
@flinkbot run travis
@flinkbot run azure

**Comment by carp84** [comment]
Thanks for the ping @pnowojski @rkhachatryan , and sorry for the late response. I've noticed this one and asked @Myasuka and @klion26 to take a look. I myself am also checking the changes but with slow progress, will drop some comments once ready.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/StreamOperatorStateHandlerTest.java [hunk 7]
```diff
@@ -76,12 +79,16 @@ public void testFailingBackendSnapshotMethod() throws Exception {
...
+			RunnableFuture<SnapshotResult<StateObjectCollection<InputChannelStateHandle>>> inputChannelStateFuture = new CancelableFuture<>();
+			RunnableFuture<SnapshotResult<StateObjectCollection<ResultSubpartitionStateHandle>>> resultSubpartitionStateFuture = new CancelableFuture<>();
 
 			OperatorSnapshotFutures operatorSnapshotResult = new OperatorSnapshotFutures(
 				keyedStateManagedFuture,
 				keyedStateRawFuture,
 				operatorStateManagedFuture,
-				operatorStateRawFuture);
+				operatorStateRawFuture,
+				inputChannelStateFuture,
```
Comment (klion26) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/StreamOperatorStateHandlerTest.java]: Do we need to test that the `inputChannelStateFuture` and `resultSubpartitionStateFuture` is cancelled below?

Another question: do we have a test to cover that `inputChannelState` or `resultSubpartitionState` failed will fail the whole checkpoint?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/util/OperatorSnapshotUtil.java [hunk 8]
```diff
@@ -79,6 +81,18 @@ public static void writeStateHandle(OperatorSubtaskState state, String path) thr
...
+			Collection<InputChannelStateHandle> inputChannelStateHandles  = state.getInputChannelState();
+			dos.writeInt(inputChannelStateHandles.size());
+			for (InputChannelStateHandle inputChannelStateHandle : inputChannelStateHandles) {
+				MetadataV3Serializer.INSTANCE.serializeInputChannelStateHandle(inputChannelStateHandle, dos);
+			}
+
+			Collection<ResultSubpartitionStateHandle> resultSubpartitionStateHandles  = state.getResultSubpartitionState();
+			dos.writeInt(inputChannelStateHandles.size());
+			for (ResultSubpartitionStateHandle resultSubpartitionStateHandle : resultSubpartitionStateHandles) {
+				MetadataV3Serializer.INSTANCE.serializeResultSubpartitionStateHandle(resultSubpartitionStateHandle, dos);
```
Comment (klion26) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/util/OperatorSnapshotUtil.java]: Do we need to update the `readStateHandle` function below.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV2V3SerializerBase.java [hunk 9]
```diff
@@ -212,46 +218,66 @@ protected MasterState deserializeMasterState(DataInputStream dis) throws IOExcep
...
 
 	protected void serializeSubtaskState(OperatorSubtaskState subtaskState, DataOutputStream dos) throws IOException {
-		final OperatorStateHandle managedOperatorState = extractSingleton(subtaskState.getManagedOperatorState());
-		if (managedOperatorState != null) {
+		serializeSingleton(subtaskState.getManagedOperatorState(), dos, MetadataV2V3SerializerBase::serializeOperatorStateHandle);
+		serializeSingleton(subtaskState.getRawOperatorState(), dos, MetadataV2V3SerializerBase::serializeOperatorStateHandle);
+		serializeKeyedStateCol(subtaskState.getManagedKeyedState(), dos);
+		serializeKeyedStateCol(subtaskState.getRawKeyedState(), dos);
+		serializeCollection(subtaskState.getInputChannelState(), dos, this::serializeInputChannelStateHandle);
+		serializeCollection(subtaskState.getResultSubpartitionState(), dos, this::serializeResultSubpartitionStateHandle);
```
Comment (Myasuka) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV2V3SerializerBase.java]: I noticed something weird here. Although we would not use `MetadataV2Serializer` to serialize checkpoint after this version, however, you actually changed the serialized format of `MetadataV2Serializer`, it would serialize two `0` into the stream here. On the other hand, `MetadataV2Serializer#deserializeSubtaskState` would ignore the written part but just return a `StateObjectCollection.empty()`.
In other words, this PR actually changed the logic of serialize and deserialize of `MetadataV2Serializer`. Although we might not meet this error in real environment as we would always use `MetadataV3Serializer` to serialize.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ChannelStateNoRescalingPartitionerTest.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+		int[] offsetSizes = {0, 1, 2};
+		final List<Function<OperatorSubtaskState, ?>> extractors = Arrays.asList(
+			OperatorSubtaskState::getInputChannelState,
+			OperatorSubtaskState::getResultSubpartitionState
+		);
+		for (int oldParallelism : parLevels) {
+			for (int newParallelism : parLevels) {
+				for (int offsetSize : offsetSizes) {
+					for (Function<OperatorSubtaskState, ?> stateExtractor : extractors) {
+						params.add(new Object[]{oldParallelism, newParallelism, offsetSize, stateExtractor});
```
Comment (Myasuka) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ChannelStateNoRescalingPartitionerTest.java]: I think there existed too many parameters. We could remove `stateExtractor`, please refer to my suggestion below.
Besides, why we need `offsetSize` here?

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ChannelStateNoRescalingPartitionerTest.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+			if (!shouldFail()) {
+				throw e;
+			} else {
+				return;
+			}
+		}
+		if (shouldFail()) {
+			fail("expected to fail for: oldParallelism=" + oldParallelism + ", newParallelism=" + newParallelism + ", offsetsSize=" + offsetsSize + ", extractState=" + extractState);
+		}
+	}
```
Comment (Myasuka) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ChannelStateNoRescalingPartitionerTest.java]: I think we could change the verify logic to 
~~~java
	@Test
	public <T extends AbstractChannelStateHandle<?>> void testNoRescaling() {
		OperatorState state = new OperatorState(OPERATOR_ID, oldParallelism, oldParallelism);
		state.putState(0, new OperatorSubtaskState(
			empty(),
			empty(),
			empty(),
			empty(),
			singleton(new InputChannelStateHandle(new InputChannelInfo(0, 0), new EmptyStreamStateHandle(), getOffset())),
			singleton(new ResultSubpartitionStateHandle(new ResultSubpartitionInfo(0, 0), new EmptyStreamStateHandle(), getOffset()))));
		verifyNoRescalingChannelState(state, OperatorSubtaskState::getInputChannelState);
		verifyNoRescalingChannelState(state, OperatorSubtaskState::getResultSubpartitionState);
	}

	private <T extends AbstractChannelStateHandle<?>> void verifyNoRescalingChannelState(OperatorState state, Function<OperatorSubtaskState, StateObjectCollection<T>> stateExtractor) {
		try {
			// noinspection unchecked
			StateAssignmentOperation.reDistributePartitionableStates(
				singletonList(state),
				newParallelism,
				singletonList(OPERATOR_ID),
				stateExtractor,
				channelStateNonRescalingRepartitioner("test"));
		} catch (IllegalArgumentException e) {
			if (!shouldFail()) {
				throw e;
			} else {
				return;
			}
		}
		if (shouldFail()) {
			fail("expected to fail for: oldParallelism=" + oldParallelism + ", newParallelism=" + newParallelism + ", offsetsSize=" + offsetsSize + ", stateExtractor=" + stateExtractor);
		}
	}
~~~

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ChannelStateNoRescalingPartitionerTest.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+
+/**
+ * Simple test for
+ * {@link org.apache.flink.runtime.checkpoint.StateAssignmentOperation#channelStateNonRescalingRepartitioner channelStateNonRescalingRepartitioner}.
+ * It checks whether partitioner fails or not using property-based approach.
+ */
+@RunWith(Parameterized.class)
+public class ChannelStateNoRescalingPartitionerTest {
+
+	@Parameters
```
Comment (Myasuka) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ChannelStateNoRescalingPartitionerTest.java]: I think current `parameters` are not so readable, how about add hint like:
~~~java
@Parameters(name = "oldParallelism: {0}, newParallelism: {1}")
~~~
I did not include `offsetSize` and `stateExtractor` here as I think we might not need them.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java [hunk 13]
```diff
@@ -2022,17 +2026,18 @@ private File createSavepointWithOperatorState(long savepointId, OperatorID... op
 	}
 
 	private Collection<OperatorState> createOperatorState(OperatorID... operatorIds) {
+		Random random = new Random(0x42);
```
[the remark this answers is not in the document]
Comment (rkhachatryan) [reply flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java]: The values from this random don't have to be globally unique. So it doesn't matter if it's the same seed or not for different calls. 
However, if this method would be used differently, it could matter. 
So I'll remove the seed.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/OperatorSnapshotFinalizerTest.java [hunk 14]
```diff
@@ -43,68 +55,72 @@
...
+		KeyedStateHandle keyedTemplate = StateHandleDummyUtil.createNewKeyedStateHandle(new KeyGroupRange(0, 0));
+		OperatorStateHandle operatorTemplate = StateHandleDummyUtil.createNewOperatorStateHandle(2, random);
+		InputChannelStateHandle inputChannelTemplate = StateHandleDummyUtil.createNewInputChannelStateHandle(2, random);
+		ResultSubpartitionStateHandle resultSubpartitionTemplate = StateHandleDummyUtil.createNewResultSubpartitionStateHandle(2, random);
+
+		SnapshotResult<KeyedStateHandle> manKeyed = withLocalState(deepDummyCopy(keyedTemplate), deepDummyCopy(keyedTemplate));
+		SnapshotResult<KeyedStateHandle> rawKeyed = withLocalState(deepDummyCopy(keyedTemplate), deepDummyCopy(keyedTemplate));
+		SnapshotResult<OperatorStateHandle> manOper = withLocalState(deepDummyCopy(operatorTemplate), deepDummyCopy(operatorTemplate));
+		SnapshotResult<OperatorStateHandle> rawOper = withLocalState(deepDummyCopy(operatorTemplate), deepDummyCopy(operatorTemplate));
+		SnapshotResult<StateObjectCollection<InputChannelStateHandle>> inputChannel = withLocalState(
```
[the remark this answers is not in the document]
Comment (rkhachatryan) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/OperatorSnapshotFinalizerTest.java]: Local recovery is not a priority. But I added it in places where local and non-local state is managed similarly so in future it would be easier to add support for it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/ChannelStateHandleSerializer.java [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+		serializeChannelStateHandle(handle, dos, (info, dataOutputStream) -> {
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/ChannelStateHandleSerializer.java]: rename `dataOutputStream` to `ignored` to not confuse it with `dos`?

**Comment by pnowojski** [comment]
I have only one nit comment remaining. @carp84 could you take care of merging this PR once it has your and/or @Myasuka LGTM?

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ChannelStateNoRescalingPartitionerTest.java (hunk 11 above)
Comment (rkhachatryan) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ChannelStateNoRescalingPartitionerTest.java]: All of the parameters are indeed parameters of the tested method.
I think it's better to keep their combinations in a single place so it's easy to see what combinations are tested.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ChannelStateNoRescalingPartitionerTest.java (hunk 10 above)
Comment (rkhachatryan) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ChannelStateNoRescalingPartitionerTest.java]: Parameters here reflect the parameters of the tested method. 
To me, this generation of all combinations with nested loops is simpler than having parameters generated in different places. I've added a comment to make it clear.

> why we need offsetSize here?

No offsets means no state so re-scaling with no offsets should be allowed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV2V3SerializerBase.java (hunk 9 above)
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV2V3SerializerBase.java]: You are right. I'll move this code to V3 subclass to keep V2 format intact.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/LocalStateForwardingTest.java [hunk 16]
```diff
@@ -232,7 +239,27 @@ public void storeLocalState(
...
+		Assert.assertEquals(snapshotResult.getJobManagerOwnedSnapshot(), jmState);
+		Assert.assertEquals(snapshotResult.getTaskLocalSnapshot(), tmState);
+	}
+
 	private static <T extends StateObject> RunnableFuture<SnapshotResult<T>> createSnapshotResult(Class<T> clazz) {
 		return DoneFuture.of(SnapshotResult.withLocalState(mock(clazz), mock(clazz)));
 	}
+
+	private static <T extends StateObject> RunnableFuture<SnapshotResult<StateObjectCollection<T>>> createSnapshotCollectionResult(Class<T> clazz) {
+		return DoneFuture.of(SnapshotResult.withLocalState(singleton(mock(clazz)), singleton(mock(clazz))));
```
[the remark this answers is not in the document]
Comment (rkhachatryan) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/LocalStateForwardingTest.java]: I think it's better to either have all the fields of a class mocked or all not mocked, but not mixed.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/util/OperatorSnapshotUtil.java (hunk 8 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/util/OperatorSnapshotUtil.java]: Yes, you're right. Thanks for pointing this out.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/StreamOperatorStateHandlerTest.java (hunk 7 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/StreamOperatorStateHandlerTest.java]: > Do we need to test that the inputChannelStateFuture and resultSubpartitionStateFuture is cancelled below?

Yes, that makes sense.

> do we have a test to cover that inputChannelState or resultSubpartitionState failed will fail the whole checkpoint?

AFAIK we don't have such tests. But there are tests that check that all the handles were passed to the CheckpointCoordinator, including the new handles. This means that TaskStateSnapshot won't be constructed if any of these futures fail.

**Comment by Myasuka** [comment]
I think this PR LGTM now.

**Comment by rkhachatryan** [comment]
Thanks for reviewing @Myasuka .

@klion26, I've addressed some issues and answered your comments, can you please take a look?

**Comment by klion26** [comment]
Thanks for the update, LGTM now.

**Comment by rkhachatryan** [comment]
@carp84 
can you please take a look at this PR?
